### PR TITLE
Data module alpha2: part 3 - lookup tables

### PR DESCRIPTION
**Pull request by aaronc**
This introduces a lookup table data structure to be used on top of the `KVStore` for efficiently storing data module IRIs.

IRIs can be somewhat long and will be referenced often in the data store so I wanted to do some upfront compression.

My initial thought was to use 64-bit auto-increment IDs for these, but I realized it would probably be more efficient to use a fast, non-cryptographic hash function (currently FNV-1a 64-bit) and just deal with collisions (which are rare) using disambiguation bytes.

I actually think this construction is pretty interesting in its own right and may have other use cases in state management. This data structure compresses keys into 4-6 bytes of storage using a single lookup table. With auto-incremented IDs, we would actually be storing 8-byte IDs (or varints) and instead of one table we would need two tables: one for the forward mapping (value -> ID) and another for the reverse (ID -> value). Seems like a noticeable improvement.

### CODE REVIEW

**Comment by aaronc** [comment]
Just a note regarding test coverage here. The other paths in this diff can be tested by increasing the number of cases tested but this makes the tests take longer. Currently this is testing 10^5 random values which does not trigger any hash collisions. To trigger collisions, we need at least 10^6 or 10^7 values and then these other paths will show up in the coverage diff. But again this leads to significantly longer testing times... Suggestions on how to approach this?

**Review by blushi — COMMENTED** [review]
For tests coverage, what about choosing very suboptimal parameters that would give more collisions? Playing with `MinLength` for instance or maybe building some trivial hash function that would give a good amount of collisions.

**Comment by aaronc** [comment]
> For tests coverage, what about choosing very suboptimal parameters that would give more collisions? Playing with `MinLength` for instance or maybe building some trivial hash function that would give a good amount of collisions.

Good idea. I built a 16-bit hash function that triggers all the cases. Decreasing `MinLength` for a semi-optimal function actually makes some edge cases harder to hit because there's more possible hash lengths for reducing collisions. Anyway, test coverage should be good now...

**Comment by robert-zaremba** [comment]
>  Suggestions on how to approach this?

Two other ideas:
* brute force to find collisions, and then use that collisions in tests (hard code instead of picking random numbers).
* hash function is customizable - so in tests we could use some simple hash function (eg taking first 4 bits - so on average we will have collision 1/16 cases).

File: x/data/server/lookup/table.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+	// Deal with collisions by appending a varint disambiguation value.
+	// Such collisions are almost impossible with good settings, but can
+	// happen with a sub-optimal hash function.
+	var i uint64 = 0
+	preLen := t.prefixLen + t.hashLen
+	for {
+		id = id[:t.bufLen]
+		n := binary.PutUvarint(id[preLen:], i)
+		id = id[:preLen+n]
```
Comment (robert-zaremba) [inline x/data/server/lookup/table.go:135]:
```suggestion
		n := binary.PutUvarint(id[preLen:], i)
		copy(id[t.bufLen:], n)
```

File: x/data/server/lookup/table.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+			return fnv.New64a()
+		}
+	}
+
+	hashLen := len(newHash().Sum(nil))
+	if minLength > hashLen {
+		return nil, fmt.Errorf("option MinLength %d is greater than hash length %d", minLength, hashLen)
+	}
+
+	bufLen := prefixLen + hashLen + binary.MaxVarintLen64
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go:52]: Also, I think `4` bytes are enough for disambiguation bytes - no need for 10 bytes - that's more then `uint64`.

File: x/data/server/lookup/table.go (hunk 1 above)
Comment (robert-zaremba) [reply x/data/server/lookup/table.go:135]: Otherwise you need to use `append(id, n...)` instead of `id = id[:preLen+n]`

File: x/data/server/lookup/table.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		store.Set(id, value)
+		return true
+	}
+
+	// id exists, check if equal
+	if bytes.Equal(value, bz) {
+		return true
+	}
+
+	return false
```
Comment (robert-zaremba) [inline x/data/server/lookup/table.go:158]:
```suggestion
	return bytes.Equal(value, bz)
```

File: x/data/server/lookup/doc.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+Shortened identifiers are created or retrieved using the idempotent Table.GetOrCreateID method.
+Identifiers will be 4 bytes long and retrieved with a single KV-store lookup in the vast majority
+of cases and will sometimes be 5 or rarely 6 bytes long and require 2 to 3 total reads. In some
+rare cases (which have not appeared in tests), identifiers may be longer or require more lookups.
+
+Given a shortened identifier, the underlying binary data can be retrieved with the Table.GetValue
+method.
+*/
+package lookup
```
Comment (robert-zaremba) [inline x/data/server/lookup/doc.go:18]: how about moving this package to other place? the package is general, but it's anchored to the server (` x/data/server/lookup/`).

File: x/data/server/lookup/table_test.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/tendermint/tendermint/libs/rand"
+)
+
+func TestTable(t *testing.T) {
+	// test default case with good params
+	table, err := NewTable(nil)
+	require.NoError(t, err)
+	testTable(t, table, 5)
```
Comment (robert-zaremba) [inline x/data/server/lookup/table_test.go:19]: how long it takes to run `1e5` tests? Maybe we can make it smaller as discussed in the PR comments?

File: x/data/server/lookup/table_test.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+type sixteenByteHash struct {
+	hash.Hash
+}
+
+func (h sixteenByteHash) Sum(b []byte) []byte {
+	bz := h.Hash.Sum(b)
+	// just return two bytes
+	return bz[:len(b)+2]
```
Comment (robert-zaremba) [inline x/data/server/lookup/table_test.go:41]:
```suggestion
	return bz[:2]
```

Comment (robert-zaremba) [reply x/data/server/lookup/table_test.go:41]: otherwise we should update the comment.

File: x/data/server/lookup/table_test.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+			return sixteenByteHash{
+				fnv.New32(),
+			}
+		},
+	})
+	require.NoError(t, err)
+	testTable(t, table, 5)
+}
+
+type sixteenByteHash struct {
```
Comment (robert-zaremba) [inline x/data/server/lookup/table_test.go:34]:
```suggestion
type sixteenBitHash struct {
```

**Review by robert-zaremba — CHANGES_REQUESTED** [review]
Added few optimizations and fix for the disambiguation bytes append.

File: x/data/server/lookup/table.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	return store.Get(id)
+}
+
+func (t table) GetOrCreateID(store KVStore, value []byte) []byte {
+	id, _ := t.getOrCreateID(store, value)
+	return id
+}
+
+func (t table) getOrCreateID(store KVStore, value []byte) (id []byte, numCollisions int) {
+	hasher := t.newHash()
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:107]: I think we do want this to be safe for concurrent usage. I didn't document it here, but one thought I've had in mind is that this sort of lookup table is much better for auto-generated IDs during parallel execution then auto-incremented `uint64`s. With auto-incremented `uint64`s there is write contention for the sequence value. Here we know that the only keys that need to be written have the prefix `hash(value)[:minLength]` so parallelism is much easier.

Of course, now that I think of it, there is also an attack where an attacker could just send a bunch of identical transactions to reduce parallelism, not sure if there's a way to mitigate that risk...

File: x/data/server/lookup/table.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:52]: In testing, I never came across a scenario where more than 6 total bytes were needed with 10^8 cases so the algorithm is overly conservative and always allocates a buffer which is too big. I do want it to be theoretically impossible to have a collision which is why the varint is appended in the end, but with any decent hash function and proper parameters it will _never_ be needed.

File: x/data/server/lookup/doc.go (hunk 4 above)
Comment (aaronc) [reply x/data/server/lookup/doc.go:18]: makes sense...

File: x/data/server/lookup/table.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+)
+
+// Table is the interface for interacting with a lookup table.
+type Table interface {
+	// GetOrCreateID is an idempotent method for creating or retrieving an unique
+	// shortened identifier for the provided binary value.
+	GetOrCreateID(store KVStore, value []byte) []byte
+
+	// GetValue returns the binary data (if any) corresponding to the provided shortened identifier.
+	GetValue(store KVStore, id []byte) []byte
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:18]: I actually created a separate `KVStore` interface with just `Get` and `Set` in this package so there's no dependency on the SDK's `KVStore`. Having the `KVStore` as a member to the `Table` doesn't work really because it's transactional and we don't want to retain a handle to invalid `KVStore`s

File: x/data/server/lookup/table.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+	prefixLen := len(options.Prefix)
+	minLength := options.MinLength
+	if minLength == 0 {
+		minLength = 4
+	}
+
+	newHash := options.NewHash
+	if newHash == nil {
+		newHash = func() hash.Hash {
+			return fnv.New64a()
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:43]: How about just blake hashes which are in the go stdlib? Is there a concern in adding a dependency on SipHash? I guess not since it's in BTC? I'm fine to use whatever is the best, most peer-reviewed hash function for this purpose.

**Comment by aaronc** [comment]
> > Suggestions on how to approach this?
> 
> Two other ideas:
> 
> * brute force to find collisions, and then use that collisions in tests (hard code instead of picking random numbers).
> * hash function is customizable - so in tests we could use some simple hash function (eg taking first 4 bits - so on average we will have collision 1/16 cases).

Yeah using just a 16-bit hash and 10^5 cases was enough to force enough collisions.

File: x/data/server/lookup/table_test.go (hunk 5 above)
Comment (aaronc) [reply x/data/server/lookup/table_test.go:19]: It's quite quick actually on my machine

File: x/data/server/lookup/table.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go:52]: Correct. I feel we should update the documentation. Now it says that the minimum ID length is `4` bytes. But it should be `prefixLen + minHashLen`.

**Comment by robert-zaremba** [comment]
> Yeah using just a 16-bit hash and 10^5 cases was enough to force enough collisions.

But it's still not deterministic? So once every X tests we will have false negative.

**Comment by aaronc** [comment]
> > Yeah using just a 16-bit hash and 10^5 cases was enough to force enough collisions.
> 
> But it's still not deterministic? So once every X tests we will have false negative.

It seems pretty reliable that it triggers the collisions with these params, there are a lot of collisions with such a small hash.

File: x/data/server/lookup/table.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+func (t table) GetOrCreateID(store KVStore, value []byte) []byte {
+	id, _ := t.getOrCreateID(store, value)
+	return id
+}
+
+func (t table) getOrCreateID(store KVStore, value []byte) (id []byte, numCollisions int) {
+	hasher := t.newHash()
+	_, err := hasher.Write(value)
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:110]: No, but I think an error here is almost impossible so changing the func signature for this seems unnecessary. In what cases could `hash.Write` error?

[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:110]: We can change the func signature to be more conservative if you think it's better

File: x/data/server/lookup/table_test.go (hunk 6 above)
Comment (aaronc) [reply x/data/server/lookup/table_test.go:41]: I'll update the comment. This is to conform to the specs of `Sum`

File: x/data/server/lookup/table.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go:43]: `SipHash` is widely used for hash map implementation. It's a small dependency. It could create a standard push back (to use only stdlib hashes). BTW, `SipHash` is designed by  Jean-Philippe (he is a prof at EPFL - neighbor university), which is an author of `blake` family functions. I can ask him - was chatting with him in the past.

File: x/data/server/lookup/table.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go:18]: Ah, right - in production it will change between the calls. Makes sense.

File: x/data/server/lookup/table.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go:110]: Right - we don't need. Didn't put enough attention that it comes from hasher. All hashers I know will never return error. The error is there to comply to `io.Writer` interface. So we can add a comment, that the error will never happen.

File: x/data/server/lookup/table.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:43]: Since this is a dependency on #236, how about we address this as a follow up PR after we've decided on the correct hash function and implementation?

I'm planning to address the other comments and then it would be good to merge so #236 can get unblocked.

File: x/data/server/lookup/table.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+	id = append(id, t.prefix...)
+
+	// take the first i bytes of hashBz starting with t.minLen and increasing
+	// in cases where there are collisions
+	for i := t.minLen; i <= t.hashLen; i++ {
+		id = append(id[t.prefixLen:], hashBz[:i]...)
+		if tryID(store, id, value) {
+			return id, i - t.minLen
+		}
+		id = id[:t.prefixLen]
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go:124]: thanks 👍 . note that `0` should be `t.bufLen + t.minLen`

**Comment by clevinson** [comment]
@robert-zaremba are these change requests blocking? If not, can you remove the change requests so we can get this merged in?

File: x/data/server/lookup/table.go (hunk 1 above)
Comment (aaronc) [reply x/data/server/lookup/table.go:135]: hmm... that doesn't quite work...

File: x/data/server/lookup/table_test.go (hunk 5 above)
Comment (aaronc) [reply x/data/server/lookup/table_test.go:19]: less than 1sec I believe

File: x/data/server/lookup/doc.go [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+A new lookup Table can be created with the NewTable() function. Advanced users can use
+the NewTableWithOptions() function to tweak the underlying parameters, but the defaults
+were chosen based on testing and should provide a good balance of performance and storage
+efficiency.
+
+Shortened identifiers are created or retrieved using the idempotent Table.GetOrCreateID method.
+Identifiers will be 4 bytes long and retrieved with a single KV-store lookup in the vast majority
+of cases and will sometimes be 5 or rarely 6 bytes long and require 2 to 3 total reads. In some
+rare cases (which have not appeared in tests), identifiers may be longer or require more lookups.
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/doc.go:13]: I added something, let me know what you think

File: x/data/server/lookup/table.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go]: yes, in the first line.

File: x/data/server/lookup/table.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go]: yes, let's just create a task.

File: x/data/server/lookup/doc.go (hunk 4 above)
Comment (aaronc) [reply x/data/server/lookup/doc.go]: move the package to `store/lookup`. what do you think?

File: x/data/server/lookup/table.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go]: so let's put in the constructor doc comment, that it's safe for parallel usage.

[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go]: hmm... makes more sense in the method docs themselves no?

[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go]: although, it isn't really safe for any concurrent usage - only synchronized parallel usage because we have no update method for this store...

File: x/data/server/lookup/table.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go]: The minimum ID length is 4 bytes, it just might be stored with a prefix in the KVStore, but that doesn't change the ID length

File: x/data/server/lookup/table_test.go (hunk 5 above)
Comment (robert-zaremba) [reply x/data/server/lookup/table_test.go]: 1 sec for a unit test it's a lot.

File: x/data/server/lookup/table.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/lookup/table.go]: https://github.com/regen-network/regen-ledger/issues/248

File: x/data/server/lookup/table.go (hunk 1 above)
Comment (aaronc) [reply x/data/server/lookup/table.go]: I left it as is because this wasn't working

File: store/lookup/table.go [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+	// in cases where there are collisions
+
+	for i := t.minLen; ; i++ {
+		if tryID(store, id, value) {
+			return id, i - t.minLen
+		}
+
+		if i >= t.hashLen {
+			break
+		}
```
Comment (robert-zaremba) [inline store/lookup/table.go:131]: the 3 lines above can be moved to the for condition, because at the start of the loop we always have `i < t.minLen`
```
for i := t.minLen; i<t.hashLen; i++ {
```

File: store/lookup/table.go [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+		id = id[:t.bufLen]
+		n := binary.PutUvarint(id[preLen:], i)
+		id = id[:preLen+n]
+		if tryID(store, id, value) {
+			return id, t.hashLen + int(i) - t.minLen
+		}
+
+		i++
+	}
+}
```
Comment (robert-zaremba) [inline store/lookup/table.go:150]:
```suggestion
	preLen := t.prefixLen + t.hashLen
	for i:=0; ; i++{
		id = id[:t.bufLen]
		n := binary.PutUvarint(id[preLen:], i)
		id = id[:preLen+n]
		if tryID(store, id, value) {
			return id, t.hashLen + int(i) - t.minLen
		}
	}
}
```

**Review by robert-zaremba — APPROVED** [review]
Left few minor suggestions. Maybe in other PR we can optimize the tests further.

File: store/lookup/table.go (hunk 14 above)
Comment (aaronc) [reply store/lookup/table.go:131]: at the start of the loop `i == t.minLen` and if `t.minLen == t.hashLen` we have array out of bounds

**Comment by aaronc** [comment]
> Left few minor suggestions. Maybe in other PR we can optimize the tests further.

If you think it's a useful algorithm, I would like to refine and optimize it further. Feels like there are some more general use cases.

File: x/data/server/lookup/doc.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/doc.go]: better, thanks.

File: x/data/server/lookup/table.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply x/data/server/lookup/table.go]: but then it depends on the store implementation. So the structure itself is safe to use in a parallel environment.

File: x/data/server/lookup/doc.go (hunk 4 above)
Comment (robert-zaremba) [reply x/data/server/lookup/doc.go]: perfect
